### ISSUE TITLE
Very slow compilation of gradient (esp. on CPU)

### ISSUE THREAD

**Issue by swertz** [google/jax#4924]
First of all, thanks for this great package! I'm trying to use it to build a tool for maximum likelihood fits in particle physics and it's a pleasure to work with.

I'm still somewhat undecided between JAX and Tensorflow, though... the issue I'm encountering with JAX is that the compilation times can become quite large. While compilation of the likelihood itself is reasonably fast (a few seconds), it's the compilation of the gradient that's really a killer (up to 10 minutes). Using Tensorflow `tf.function`s (without JITting) the evaluation itself is quite a bit slower, but the tracing of the gradient is faster.

In the end, I wouldn't mind a somewhat slower (less optimized) evaluation of the likelihood/gradient (even by an order of magnitude!) if it means gaining an order of magnitude in the compilation time, since in practice what the user does with those functions after compiling them would rarely take more than a few minutes.

What's also strange is that compilation is much faster for a GPU backend than for CPU (in the example below, 40s vs. 5 minutes). In practice, we would more often work on CPU than GPU (latency costs are usually too high on GPU for this kind of application, in particular since we still use non-JAX `scipy.optimize` minimization algorithms).

Here are links to a self-contained example:
[colab on CPU](https://colab.research.google.com/drive/1QofCXt902lYrKhKuTnCr2uzYU-3YABn1?usp=sharing)
[colab on GPU](https://colab.research.google.com/drive/1ihvf6YtHzY8m5uyiceamjDvEDgsAPfKf?usp=sharing)

I hope they are sufficiently self-explanatory...! I've tried to make them as short as possible while still retaining the computational complexity of the full package.
You'll note that there are two nested python `for` loops in the compiled function, that would be unrolled by JAX, but I most likely cannot avoid that in my tool to keep the freedom for user-supplied inputs. However, I'm not sure these are the culprit for the slow compiles, since the compilation of `pred` and `nll` is fast and unrolling also happens there...

I also attach an [XLA dump](https://github.com/google/jax/files/5553264/hlo_dump.tar.gz) for the CPU case.

**Comment by shoyer** [comment]
I think this FAQ entry applies to your code example, "jit decorated function is very slow to compile":
https://jax.readthedocs.io/en/latest/faq.html#jit-decorated-function-is-very-slow-to-compile

Gradient calculations can be a good bit more complicated than forward calculations, which I guess could explain the compile time difference. For example, `jax.make_jaxpr(nll)(x0, nObs)` is 3943 lines vs 9725 lines for `jax.make_jaxpr(nll_grad)(x0, nObs)`. Note that XLA compile time generally grows super-linearly in the size of the inputs.

**Comment by mattjj** [comment]
Thanks for the report!

There's likely a lot of headroom for XLA:CPU to improve compile times, so we could forward this along to the XLA team as an example.

Another recourse is just to use `jit` on smaller subfunctions.

But ultimately we need to improve XLA:CPU compile times.

**Comment by mattjj** [comment]
There are some XLA:CPU improvements in the longer-term pipeline that can probably speed up the XLA:CPU compile time to be on par with the XLA:GPU compile time (thanks for taking a look at this, @d0k !). But ultimately this kind of unrolling really exacerbates the issue:

```python
@jit
def pred(x):
    yields = jnp.zeros(template.shape)
    for j in range(nProc):
        factors = jnp.ones(template.shape)
        for idxs,fn in fns:
            factors = fn(x[idxs]) * factors # Here the functions are all similar and known in advance, but in practice they are user-supplied and arbitrary
        p_yields = factors * (template+j) # In practice, the 'template' would be user-supplied and different for each j
        yields += p_yields
    return yields
```

You might consider removing `jit` from there (and `nll_grad`), and perhaps just jitting `fns`. That would certainly reduce the compile time, but likely at great cost to the execution time. Still, for the time being, that's probably the best way to trade off compile time and execution time!

**Comment by swertz** [comment]
Thanks a lot for taking a look!

I've tried to `jit` the subfunctions only and compilation time indeed reduces by a factor 2-10 (depending on the configuration) and scales more linearly with the parameters (e.g. `nProc` in the example linked), however the evaluation of `nll` becomes around 100x slower and `nll_grad` becomes 500-1000x slower, which is a little too much to swallow... It seems XLA is quite effective at optimizing things indeed!

> Gradient calculations can be a good bit more complicated than forward calculations, which I guess could explain the compile time difference. For example, `jax.make_jaxpr(nll)(x0, nObs)` is 3943 lines vs 9725 lines for `jax.make_jaxpr(nll_grad)(x0, nObs)`. Note that XLA compile time generally grows super-linearly in the size of the inputs.

I suppose there is no way to get XLA to be *a bit* less aggressive with its optimizations, to get a linear scaling with the program size instead of the quadratic scaling I observe as a function of `nProc`? (very naively, in the spirit of `gcc`'s `-O1`, `-O2`, `-O3`)

In the sense that a "classical" compiler would compile files or libraries separately and finally link them together in an executable, would it ever be possible to compile sub-functions (as `fn` here) separately and "link" those when compiling `pred`, `nll` or `nll_grad`? This would not bring any optimizations (e.g. from fusing operations) but would avoid all the python overhead in the larger functions...

Getting back to my example, are the unrolled `for` loops really the only culprits for the large resulting gradient code, or are there other things that could/should be expressed differently? I'm thinking e.g. about those indexing/reshaping operations in `pred` and `fn` (sorry, reading the generated jaxpr is a little though for the average user!):
```
factors = fn(x[idxs]) * factors
```
```
x = x[x_idxs]
x = jnp.reshape(x, x_shape)
```

**Comment by swertz** [comment]
I have managed to express the function differently to avoid to avoid the explicit `for` loops, and compilation time is now reduced below 10 seconds for all practical cases :)

That required however being quite aggressive in padding the input arrays, so memory usage has become more of an issue... Is there any plan to support sparse tensors in the future?

**Comment by mattjj** [comment]
Glad to hear you found a solution!

> I suppose there is no way to get XLA to be a bit less aggressive with its optimizations, to get a linear scaling with the program size instead of the quadratic scaling I observe as a function of nProc?

There's not a robust way to do this that I know of, but I expect XLA:CPU to get improved in a lot of ways over the coming months, including faster compile times.

> In the sense that a "classical" compiler would compile files separately and finally link them together in an executable or library, would it ever be possible to compile sub-functions (as fn here) separately and "link" those when compiling pred, nll or nll_grad?

This is a good idea, and an ongoing discussion topic we have with the XLA team. We don't have anything concrete here yet though.

> Is there any plan to support sparse tensors in the future?

We have some experiments in the direction of sparse arrays, but nothing really concrete yet.


Since this particular issue seems to be sorted out, and since we have a lot of open issues, I'm going to close this thread for now. But please don't hesitate to open new issues!

**Comment by swertz** [comment]
Thanks again for the feedback!

**Comment by carlosgmartin** [comment]
@mattjj As of now, is there any way to get more fine-grained control over XLA compilation and the optimizations it performs, to avoid very long compile times for complicated functions? Are there any plans to add something like that?
